0011-CIR-trust-distribution.md

```diff
@@ -0,0 +1,116 @@
+Similarly for example, asset SHA hashes should be downloaded from a separate source then the assert server themselves.
+Decoupling these roles also has the benefit of improving separation of concerns between responsibilities from the identity issuer, and the trust distributor.
+
+Trust distribution will be conducted by the Operator and written to the ConfigMap `dapr-root-ca.crt` in all Namespaces.
+The Operator is a natural fit as it is not Sentry (the identity issuance server), and machinery for Kubernetes controllers already exists in the Operator today.
+ConfigMaps are a natural choice as they can be mounted by Pods & containers in Kubernetes, and trust bundles are not secrets so Secrets are not appropriate.
+There is also prior art to other projects distributing trust in this way, such as [Istio](https://github.com/istio/istio/blob/4c65649a9b116584281fadcaf8c3dd6b42d34036/istioctl/pkg/workload/workload_test.go#L340) and cert-manager's [trust-manager](https://github.com/cert-manager/trust-manager#example-bundle).
+We can also add support for writing to Kubernetes [ClusterTrustBundles](https://github.com/kubernetes/enhancements/issues/3257), though this resource is very new, and will not be available in all target Kubernetes cluster versions.
+The `dapr-root-ca.crt` ConfigMap name is consistent with Kubernetes and Istio naming.
+The operator will source the root of trust to be distributed from the mounted `dapr-root-ca.crt` Secret in the control plane namespace.
+The operator will watch for this mounted file for updates using [fsnotify](https://github.com/fsnotify/fsnotify), and distribute the contents to the named ConfigMap in all namespaces.
+
+Once propagated, the Injector, Placement, and Dapr sidecars can mount this ConfigMap and use it as the root of trust when connecting to peers.
+Similarly, when the file is updated, these services can update their local trust stores to use the new version of the bundle.
+
+The Operator will need to metadata watch all Namespaces and ConfigMaps in the cluster.
+The Operator should no fully inform these resources as that will massively increase the memory consumption of the Operator.
+In the event of a Namespace being created, the Operator will write the ConfigMap to that namespace.
+The Operator will also ensure that the `dapr-root-ca.crt` ConfigMap stays consistent with its local trust bundle version.
+
+### CA Rotation
+
+CA rotation can now be solved by the new trust bundle being _appended_ to the existing `dapr-root-ca.crt` Secret in the control plane namespace.
+This new bundle containing the old and new CA will be propagated to all services by the Operator, allowing for a zero downtime & graceful roll over of the CA.
+The Dapr CLI will be updated to automate this task, ensuring that the new appended trust bundle has been correctly propergated to all namespaces before writing the new CA to sentry.
```

Comment:
Rather than the Dapr CLI, can we get this done by an API in Sentry (that is invoked for example by deploying a new YAML)? Also due to the fact that it can take a few mins.

Reply:
Yes, I think this would be a good feature- though we need to update the CLI to perform the migration as we are supporting it already today. Or are you suggesting the CLI calls the _sentry_ API?

Reply:
In _some way_.

I understand that having the CLI invoking a Sentry API may be a problem on its own. So that's why I mentioned perhaps the CLI can deploy a YAML that triggers the method in Sentry (via a reconciler)

Reply:
I hear you. I'm dubious about creating dedicated CRD for with the overhead that that entails or overloading an existing CRD for this purpose. Perhaps we move this to a day 2 problem?

Reply:
What concerns me with saying "day 2 problem" is that the root cert rotation in Dapr has always been a "day 2 problem"... and that's where we are now :)

Reply:
True, though I suppose that at least this has a clear path forward to a concrete solution. Another point is that Sentry will not be the holder of the root certificate when using an external CA so this API is only usable in the "default" scenario, as well as not being usable in self-hosted mode.